### ISSUE TITLE
Compaction Failures Similar to #2941 on Upgrade From 2.x to 3.1.1

### ISSUE THREAD

**Issue by bdoyle0182** [apache/couchdb#3292]
## Description

we're seeing a similar issue to https://github.com/apache/couchdb/issues/2941 with random compactions on shards when upgrading from 2.x to 3.1.1. But might be completely unrelated. The compaction metadata file blew up to about 500gb over 24 hours for a shard that is about 30gb constantly hitting this error. Similarly we're also seeing large disk io on the nodes this is happening on versus nodes this is not happening on like in #2941. I've deleted the compaction files like discussed in the previous issue and seems to be working fine now, the compaction is running and the errors have stopped. disk io has gone back down.

<0.5181.0> -------- exit for compaction of ["shards/60000000-7fffffff/core_activations.1589336264"]: {badarith,[{couch_file,get_pread_locnum,3,[{file,"src/couch_file.erl"},{line,730}]},{lists,map,2,[{file,"lists.erl"},{line,1239}]},{lists,map,2,[{file,"lists.erl"},{line,1239}]},{couch_file,read_multi_raw_iolists_int,2,[{file,"src/couch_file.erl"},{line,719}]},{couch_file,handle_call,3,[{file,"src/couch_file.erl"},{line,507}]},{gen_server,try_handle_call,4,[{file,"gen_server.erl"},{line,636}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,665}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}

-------- CRASH REPORT Process (<0.32443.1691>) with 3 neighbors crashed with reason: bad arithmetic expression at couch_file:get_pread_locnum/3(line:730) <= lists:map/2(line:1239) <= couch_file:read_multi_raw_iolists_int/2(line:719) <= couch_file:handle_call/3(line:507) <= gen_server:try_handle_call/4(line:636) <= gen_server:handle_msg/6(line:665) <= proc_lib:init_p_do_apply/3(line:247); initial_call: {couch_file,init,['Argument__1']}, ancestors: [<0.3352.1692>], message_queue_len: 0, messages: [], links: [<0.3352.1692>], dictionary: [{couch_file_fd,{{file_descriptor,prim_file,{#Port<0.1924847>,92}},"..."}},...], trap_exit: false, status: running, heap_size: 28690, stack_size: 27, reductions: 13483

## Steps to Reproduce

Upgrade to 3.1.1 from 2.x mid compaction

## Expected Behaviour

Compaction to complete as expected

## Your Environment

If any specific environment details would be helpful just let me know

* CouchDB version used: 3.1.1
* Browser name and version:
* Operating system and version: centOS

**Comment by wohali** [comment]
Are you able to test the fix that was merged in https://github.com/apache/couchdb/pull/3001 ?

**Comment by bdoyle0182** [comment]
I was under the impression the fix was included in 3.1.1 which we directly upgraded to from 2.x according to the release notes under bug fixes
https://docs.couchdb.org/en/latest/whatsnew/3.1.html

**Comment by wohali** [comment]
Yup, you're right, my mistake.

@nickva Does this look like anything that #3001 would have missed?

**Comment by nickva** [comment]
This looks like a different failure from #3001. 

@bdoyle0182 what version of 2.x was running on the old nodes? What Erlang VM version, OS (which version of CentOS), and file system was used?

Looking through the issues so far it's first instance of `badarith` error from that part of the code I think.

**Comment by nickva** [comment]
Looking at the code the crash is coming from https://github.com/apache/couchdb/blob/23b98345f0db38214eaad33b159cf10dd9deaf72/src/couch/src/couch_file.erl#L729-L730

It looks like `Pos` (position) there is not an integer but something like `undefined` or `eof`.

That gets called from `couch_file:pread_terms` which is only called from the compactor's `copy_meta_data`: https://github.com/apache/couchdb/blob/23b98345f0db38214eaad33b159cf10dd9deaf72/src/couch/src/couch_bt_engine_compactor.erl#L554-L555

or compactor's `merge_doc_ids` https://github.com/apache/couchdb/blob/23b98345f0db38214eaad33b159cf10dd9deaf72/src/couch/src/couch_bt_engine_compactor.erl#L662-L669

**Comment by bdoyle0182** [comment]
Current couchdb: 3.1.1
Previous couchdb: 2.3.1

Current erlang: 20.3.8.24
Previous erlang: 19.2

centos: CentOS Linux release 7.9.2009

**Comment by bdoyle0182** [comment]
Just to update the compaction did successfully complete after being re-triggered so just deleting the compaction files seems like a fine remediation.

**Comment by nickva** [comment]
@bdoyle0182 makes sense, thanks for confirming. This probably is an upgrade issue, we have upgrade code to handle the main .couch files but not necessarily the .compact.* files I think.

@davisp I wonder if we can automatically detect the upgrade scenario and auto-delete or at least ignore the older compaction files when the format is upgraded?

We should update the docs to advise users to complete compactions on the 2.x nodes before they are upgraded to 3.x, or alternatively to delete the .compact.meta and .compact.data after the upgrade.
